drivers/rtc/rtc_bq32002.c

```diff
@@ -0,0 +1,376 @@
+#define BQ32002_SF_KEY_1 0x5E
+#define BQ32002_SF_KEY_2 0xC7
+
+/* BQ32002 counts weekdays from 1 to 7 */
+#define BQ32002_DAY_OFFSET -1
+
+/* BQ32002 counts months from 1 to 12 */
+#define BQ32002_MONTH_OFFSET -1
+
+/* Offset between first tm_year and first BQ32002 year */
+#define BQ32002_YEAR_OFFSET (2000 - 1900)
+
+/* Calibration constants, see BQ32002 datasheet, Table 12, p.16 */
+#define BQ32002_CAL_PPB_PER_LSB_POS 2034 /* 1e9 / 491520 */
+#define BQ32002_CAL_PPB_PER_LSB_NEG 4069 /* 1e9 / 245760 */
+#define BQ32002_CAL_PPB_MIN         (-31 * BQ32002_CAL_PPB_PER_LSB_POS)
+#define BQ32002_CAL_PPB_MAX         (31 * BQ32002_CAL_PPB_PER_LSB_NEG)
+
+/* IRQ frequency property enum values */
+#define BQ32002_IRQ_FREQ_ENUM_1HZ      0
+#define BQ32002_IRQ_FREQ_ENUM_512HZ    1
+#define BQ32002_IRQ_FREQ_ENUM_DISABLED 2
+
+/* RTC time fields supported by BQ32002 */
+#define BQ32002_RTC_TIME_MASK                                                                      \
+	(RTC_ALARM_TIME_MASK_SECOND | RTC_ALARM_TIME_MASK_MINUTE | RTC_ALARM_TIME_MASK_HOUR |      \
+	 RTC_ALARM_TIME_MASK_MONTH | RTC_ALARM_TIME_MASK_MONTHDAY | RTC_ALARM_TIME_MASK_YEAR |     \
+	 RTC_ALARM_TIME_MASK_WEEKDAY)
+
+struct bq32002_config {
+	struct i2c_dt_spec i2c;
+	uint8_t irq_freq;
+};
+
+struct bq32002_data {
+	struct k_sem lock;
+};
+
+static void bq32002_lock_sem(const struct device *dev)
+{
+	struct bq32002_data *data = dev->data;
+
+	(void)k_sem_take(&data->lock, K_FOREVER);
+}
+
+static void bq32002_unlock_sem(const struct device *dev)
+{
+	struct bq32002_data *data = dev->data;
+
+	k_sem_give(&data->lock);
+}
+
+static int bq32002_set_irq_frequency(const struct device *dev)
+{
+	const struct bq32002_config *config = dev->config;
+	uint8_t sf_regs[3];
+	uint8_t cfg1_val;
+	uint8_t cfg2_val;
+	int err;
+
+	switch (config->irq_freq) {
+	case BQ32002_IRQ_FREQ_ENUM_1HZ:
+		cfg1_val = BQ32002_FREQ_TEST_MASK;
+		cfg2_val = BQ32002_FTF_MASK;
+		break;
+	case BQ32002_IRQ_FREQ_ENUM_512HZ:
+		cfg1_val = BQ32002_FREQ_TEST_MASK;
+		cfg2_val = 0;
+		break;
+	default:
+		cfg1_val = BQ32002_OUT_MASK;
+		cfg2_val = 0;
+		break;
+	}
+
+	err = i2c_reg_update_byte_dt(&config->i2c, BQ32002_CAL_CFG1_REG, BQ32002_FREQ_TEST_MASK,
+				     cfg1_val);
+	if (err) {
+		return err;
+	}
+
+	/* Update FTF value if frequency output enabled */
+	if (cfg1_val & BQ32002_FREQ_TEST_MASK) {
+		sf_regs[0] = BQ32002_SF_KEY_1;
+		sf_regs[1] = BQ32002_SF_KEY_2;
+		sf_regs[2] = cfg2_val;
+		err = i2c_burst_write_dt(&config->i2c, BQ32002_SF_KEY_1_REG, sf_regs,
+					 sizeof(sf_regs));
+		if (err) {
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static int bq32002_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	const struct bq32002_config *config = dev->config;
+	int err;
+	uint8_t regs[7];
+
+	if ((timeptr == NULL) || !rtc_utils_validate_rtc_time(timeptr, BQ32002_RTC_TIME_MASK) ||
+	    (timeptr->tm_year < BQ32002_YEAR_OFFSET)) {
+		return -EINVAL;
+	}
+
+	bq32002_lock_sem(dev);
+
+	do {
```

Comment:
Use goto statements to jump, not sure if its a rule but it isthe usual way its done in tree :)

Reply:
Done